tests/PrefixCacheBootstrapperTest.php

```diff
@@ -0,0 +1,217 @@
+    $tenant2 = Tenant::create();
+    tenancy()->initialize($tenant1);
+
+    cache(['foo' => 'bar'], 10);
+    expect(cache('foo'))->toBe('bar');
+
+    tenancy()->initialize($tenant2);
+    tenancy()->end();
+
+    tenancy()->initialize($tenant1);
+    expect(cache('foo'))->toBe('bar');
+});
+
+test('prefix separate cache well enough', function () {
+    $tenant1 = Tenant::create();
+    tenancy()->initialize($tenant1);
+
+    cache()->put('foo', 'bar', 1);
+    expect(cache()->get('foo'))->toBe('bar');
+
+    $tenant2 = Tenant::create();
+    tenancy()->initialize($tenant2);
+
+    pest()->assertNotSame('bar', cache()->get('foo'));
+
+    cache()->put('foo', 'xyz', 1);
+    expect(cache()->get('foo'))->toBe('xyz');
+});
+
+test('central cache is persisted', function () {
+    cache()->put('key', 'central');
+
+    $tenant1 = Tenant::create();
+    tenancy()->initialize($tenant1);
+
+    cache()->put('key', 'tenant');
+
+    expect(cache()->get('key'))->toBe('tenant');
+
+    tenancy()->end();
+    cache()->put('key2', 'central-two');
+
+    expect(cache()->get('key'))->toBe('central');
+    expect(cache()->get('key2'))->toBe('central-two');
+});
+
+test('cache base prefix is customizable', function () {
+    $originalPrefix = config('cache.prefix');
+    $prefixBase = 'custom_';
+
+    config([
+        'tenancy.cache.prefix_base' => $prefixBase
+    ]);
+
+    $tenant1 = Tenant::create();
+    tenancy()->initialize($tenant1);
+
+    expect($originalPrefix . $prefixBase . $tenant1->getTenantKey() . ':')
+        ->toBe(app('cache')->getPrefix())
+        ->toBe(app('cache.store')->getPrefix());
+});
+
+test('prefix separate cache well enough using CacheManager dependency injection', function () {
+    $this->app->singleton(CacheService::class);
+
+    app()->make(CacheService::class)->handle();
+
+    expect(cache('key'))->toBe('central-value');
+
+    $tenant1 = Tenant::create();
+    $tenant2 = Tenant::create();
+    tenancy()->initialize($tenant1);
+
+    expect(cache('key'))->toBeNull();
+    app()->make(CacheService::class)->handle();
+    expect(cache('key'))->toBe($tenant1->getTenantKey());
+
+    tenancy()->initialize($tenant2);
+
+    expect(cache('key'))->toBeNull();
+    app()->make(CacheService::class)->handle();
+    expect(cache('key'))->toBe($tenant2->getTenantKey());
+
+    tenancy()->end();
+
+    expect(cache('key'))->toBe('central-value');
+});
+
+test('stores other than the default one are not prefixed', function () {
+    config(['cache.default' => 'redis']);
+    config(['cache.stores.redis2' => config('cache.stores.redis')]);
+
+    app(CacheManager::class)->extend('redis2', function($config) {
+        $redis = $this->app['redis'];
+
+        $connection = $config['connection'] ?? 'default';
+
+        $store = new RedisStore($redis, $this->getPrefix($config), $connection);
+
+        return $this->repository(
+            $store->setLockConnection($config['lock_connection'] ?? $connection)
+        );
+    });
```

Comment:
Is this part needed? Setting the config should be all that's needed.

Reply:
This is needed because CacheManager doesn't have a method for creating a `redis2` repository (see [this](https://github.com/laravel/framework/blob/9.x/src/Illuminate/Cache/CacheManager.php#L102-L105) – CacheManager would have to have a `createRedis2Driver`).

Also if you mean setting the `cache.default` config, the cache would be prefixed – unless we make the tenant-aware caches configurable. I'll make them configurable now

Reply:
The cache config looks like this:

<img width="418" alt="CleanShot 2023-01-05 at 2 53 32@2x" src="https://user-images.githubusercontent.com/33033094/210796071-344df833-3835-4e1f-a720-108654950394.png">

Can't we create a `redis2` store that uses a `redis` driver?

Reply:
I see, that should work

Reply:
Yeah you were right, it works without the code